File: lib/trends/steam.ts
// Steam — uses Steam Web API + storefront search. STEAM_API_KEY is optional;
// the GetNumberOfCurrentPlayers and storesearch endpoints work without it.

import { wrap, fetchJson, fmtNum } from './_helpers'
import type { TrendQuery, TrendResult, TrendDatapoint } from './types'

type StoreSearchItem = { id: number; name: string }
type StoreSearchResponse = { items: StoreSearchItem[]; total: number }
type PlayerCountResponse = { response: { player_count?: number; result?: number } }
type TopGamesResponse = { response: { ranks: Array<{ rank: number; appid: number; concurrent_in_game: number; peak_in_game: number }> } }
type AppDetailsResponse = Record<string, { success: boolean; data?: { name: string } }>

export async function fetchSteam(query: TrendQuery): Promise<TrendResult> {
  return wrap('steam', query, async () => {
    const titles = query.gameTitles ?? []

    if (titles.length === 0) {
      // No specific titles → return Steam top games by current player count.
      const top = await fetchJson<TopGamesResponse>(
        'https://api.steampowered.com/ISteamChartsService/GetMostPlayedGames/v1/',
      )
      const ranks = (top.response?.ranks ?? []).slice(0, 10)
      if (ranks.length === 0) throw new Error('empty rank list')

      // Resolve appids → names in one batch via appdetails (sequential to avoid rate limit).
      const names: Record<number, string> = {}
      for (const r of ranks.slice(0, 5)) {
        try {
          const ad = await fetchJson<AppDetailsResponse>(
            `https://store.steampowered.com/api/appdetails?appids=${r.appid}&filters=basic&cc=us`,
          )
          names[r.appid] = ad[String(r.appid)]?.data?.name ?? `appid ${r.appid}`
        } catch {
          names[r.appid] = `appid ${r.appid}`
        }
      }

      const datapoints: TrendDatapoint[] = ranks.map(r => ({
        label: names[r.appid] ?? `appid ${r.appid}`,
        value: r.concurrent_in_game,
        meta: { peak: r.peak_in_game, rank: r.rank },
      }))
      const summary = `Steam 当前在玩量 Top 10，第一名 ${names[ranks[0].appid] ?? 'unknown'} 当前 ${fmtNum(ranks[0].concurrent_in_game)} 人在玩。`
      const digest = [
        '# Steam · 当前在玩量 Top 10',
        '',
        ...ranks.map((r, i) => `${i + 1}. **${names[r.appid] ?? `appid ${r.appid}`}** · 当前 ${fmtNum(r.concurrent_in_game)} · 24h 峰值 ${fmtNum(r.peak_in_game)}`),
      ].join('\n')
      return { summary, digest, datapoints }
    }

    // Specific titles → resolve each to appid, fetch current player count.
    const results: Array<{ title: string; name: string; appid: number; players: number }> = []
    for (const t of titles.slice(0, 5)) {
      try {
        const search = await fetchJson<StoreSearchResponse>(
          `https://store.steampowered.com/api/storesearch?term=${encodeURIComponent(t)}&cc=us`,
        )
        const top = search.items?.[0]
        if (!top) continue
        const pc = await fetchJson<PlayerCountResponse>(
          `https://api.steampowered.com/ISteamUserStats/GetNumberOfCurrentPlayers/v1/?appid=${top.id}`,
        )
        const players = pc.response?.player_count ?? 0
        results.push({ title: t, name: top.name, appid: top.id, players })
      } catch {
        // skip this title; one failure doesn't sink the rest
      }
    }
    if (results.length === 0) throw new Error('no titles resolved')

    const datapoints: TrendDatapoint[] = results.map(r => ({ label: r.name, value: r.players, meta: { appid: r.appid } }))
    const summary = `Steam 当前在玩量：${results.map(r => `${r.name} ${fmtNum(r.players)}`).join('；')}。`
    const digest = [
      '# Steam · 指定游戏在玩量',
      '',
      ...results.map(r => `- **${r.name}** (appid ${r.appid}) · 当前在玩 ${fmtNum(r.players)} 人`),
    ].join('\n')
    return { summary, digest, datapoints }
  })
}

File: server/stratsquad/trends/steam.py
"""Steam — Steam Web API + storefront search. No key needed for the endpoints we use."""
from __future__ import annotations

from ..types import TrendQuery, TrendDatapoint, TrendResult
from .base import wrap, fetch_json, fmt_num


async def fetch_steam(query: TrendQuery) -> TrendResult:
    async def body() -> dict:
        titles = query.game_titles or []
        if not titles:
            top = await fetch_json(
                "https://api.steampowered.com/ISteamChartsService/GetMostPlayedGames/v1/"
            )
            ranks = (top.get("response") or {}).get("ranks", [])[:10]
            if not ranks:
                raise ValueError("empty rank list")
            names: dict[int, str] = {}
            for r in ranks[:5]:
                try:
                    ad = await fetch_json(
                        f"https://store.steampowered.com/api/appdetails?appids={r['appid']}&filters=basic&cc=us"
                    )
                    names[r["appid"]] = (ad.get(str(r["appid"])) or {}).get("data", {}).get("name", f"appid {r['appid']}")
                except Exception:
                    names[r["appid"]] = f"appid {r['appid']}"
            datapoints = [
                TrendDatapoint(
                    label=names.get(r["appid"], f"appid {r['appid']}"),
                    value=float(r["concurrent_in_game"]),
                    meta={"peak": r["peak_in_game"], "rank": r["rank"]},
                )
                for r in ranks
            ]
            first = names.get(ranks[0]["appid"], "unknown")
            summary = f"Steam 当前在玩量 Top 10，第一名 {first} 当前 {fmt_num(ranks[0]['concurrent_in_game'])} 人在玩。"
            digest_lines = ["# Steam · 当前在玩量 Top 10", ""]
            for i, r in enumerate(ranks):
                digest_lines.append(
                    f"{i+1}. **{names.get(r['appid'], 'appid '+str(r['appid']))}** · 当前 {fmt_num(r['concurrent_in_game'])} · 24h 峰值 {fmt_num(r['peak_in_game'])}"
                )
            return {"summary": summary, "digest": "\n".join(digest_lines), "datapoints": datapoints}

        results = []
        for t in titles[:5]:
            try:
                search = await fetch_json(
                    f"https://store.steampowered.com/api/storesearch?term={t}&cc=us"
                )
                top = (search.get("items") or [None])[0]
                if not top:
                    continue
                pc = await fetch_json(
                    f"https://api.steampowered.com/ISteamUserStats/GetNumberOfCurrentPlayers/v1/?appid={top['id']}"
                )
                players = ((pc.get("response") or {}).get("player_count")) or 0
                results.append({"title": t, "name": top["name"], "appid": top["id"], "players": players})
            except Exception:
                continue
        if not results:
            raise ValueError("no titles resolved")
        datapoints = [
            TrendDatapoint(label=r["name"], value=float(r["players"]), meta={"appid": r["appid"]})
            for r in results
        ]
        summary = "Steam 当前在玩量：" + "；".join(f"{r['name']} {fmt_num(r['players'])}" for r in results) + "。"
        digest = "# Steam · 指定游戏在玩量\n\n" + "\n".join(
            f"- **{r['name']}** (appid {r['appid']}) · 当前在玩 {fmt_num(r['players'])} 人" for r in results
        )
        return {"summary": summary, "digest": digest, "datapoints": datapoints}

    return await wrap("steam", query, body)
